x86: memory management adjustments

tyrantlizard77 opened the pull request:
- The memory map for ia32 SOC was using the first megabyte. This has reserved memory regions in it, relocate the start of memory for ia32 to 1MB.
- Generate page tables at runtime. Delete old buggy MMU table generation script
- qemu_x86 config adjustments
- Enable very early console for qemu_x86_coverage
- X86 now handles all its arch-specific initialization tasks before z_cstart() is called.

Please see individual commit messages for details.

Fixes: #17783
Fixes: #17782 
Fixes: #17784

Commit:
x86: generate page tables at runtime

Removes very complex boot-time generation of page tables
with a much simpler runtime generation of them at bootup.

For those x86 boards that enable the MMU in the defconfig,
set the number of page pool pages appropriately.

The MMU_RUNTIME_* flags have been removed. They were an
artifact of the old page table generation and did not
correspond to any hardware state.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86: ia32: do not use the first megabyte

After witnessing some strange errors with memory not being
what it should be, lifiting everything above 1MB has solved
it. The Zephyr binary was being loaded into memory containing
reserved regions, resulting in data corruption.

We still simulate XIP for testing purposes by setting up the
memory map as follows:

0x000000 - 0x0FFFFF : Non-present
0x100000 - 0x4FFFFF : "Flash" ROM region
0x500000 - 0x8FFFFF : "SRAM" RAM region

For a total of 9 megabytes of physical RAM used.

Fixes problems observed in some large tests when code coverage
is enabled (which increases the amount of RAM used even more).

Fixes: #17782

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
@andyross @gnuless initially I tried lifting above just the first megabyte, but still ran into issues. Moving above 2MB seems to have resolved it. I wasn't expecting any reserved memory regions between 1MB-2MB, maybe some quirk of QEMU?

andrewboie commented:
NVS fs test is failing, not entirely sure how to fix

```
c[?7l[2J[0mSeaBIOS (version rel-1.12.0-0-ga698c8995f-prebuilt.qemu.org)
Booting from ROM..***** Booting Zephyr OS build zephyr-v1.14.0-2843-g72fb988a685e *****
Running test suite test_nvs
===================================================================
starting test - test_nvs_init

    Assertion failed at ZEPHYR_BASE/tests/subsys/fs/nvs/src/main.c:72: test_nvs_init: (err == 0 is false)
Unable to get page info: -22
FAIL - test_nvs_init
===================================================================
starting test - test_nvs_write

    Assertion failed at ZEPHYR_BASE/tests/subsys/fs/nvs/src/main.c:111: test_nvs_write: (err == 0 is false)
nvs_init call failure: -22
FAIL - test_nvs_write
===================================================================
starting test - test_nvs_corrupted_write

    Assertion failed at ZEPHYR_BASE/tests/subsys/fs/nvs/src/main.c:151: test_nvs_corrupted_write: (err == 0 is false)
nvs_init call failure: -22
FAIL - test_nvs_corrupted_write
===================================================================
starting test - test_nvs_gc

    Assertion failed at ZEPHYR_BASE/tests/subsys/fs/nvs/src/main.c:221: test_nvs_gc: (err == 0 is false)
nvs_init call failure: -22
FAIL - test_nvs_gc
===================================================================
starting test - test_nvs_gc_3sectors

    Assertion failed at ZEPHYR_BASE/tests/subsys/fs/nvs/src/main.c:46: setup: (err == 0 is false)
nvs_clear call failure: -13
FAIL - test_nvs_gc_3sectors
===================================================================
starting test - test_nvs_corrupted_sector_close_operation

    Assertion failed at ZEPHYR_BASE/tests/subsys/fs/nvs/src/main.c:46: setup: (err == 0 is false)
nvs_clear call failure: -13
FAIL - test_nvs_corrupted_sector_close_operation
===================================================================
Test suite test_nvs failed.
===================================================================
PROJECT EXECUTION FAILED
[00:00:00.000,000] [1;31m<err> fs_nvs: Unable to get page info[0m
[00:00:00.000,000] [1;31m<err> fs_nvs: Unable to get page info[0m
[00:00:00.000,000] [1;31m<err> fs_nvs: Unable to get page info[0m
[00:00:00.000,000] [1;31m<err> fs_nvs: NVS not initialized[0m
[00:00:00.000,000] [1;31m<err> fs_nvs: NVS not initialized[0m
```

andrewboie commented:
The first megabyte should be enough. I wrote a test program to probe writes in the first 2MB and found these address ranges would not have any effect if writes were done to them:

0x000a0000 -> 0x000b8000
0x000c0000 -> 0x000ca000
0x000cd000 -> 0x000e8000
0x000f0000 -> 0x00100000

0x00100000 and above should be fine.

Commit:
x86: add prep_c function

Assembly language start code will enter here, which sets up
early kernel initialization and then calls z_cstart() when
finished.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
I combined this with my other patch which generates page tables at runtime, as lifting to 1MB exposed bugs in the old page table generation script, and I'd rather just remove it.

andrewboie commented on arch/x86/core/ia32/x86_mmu.c:
```diff
@@ -344,6 +364,173 @@ void z_x86_mmu_set_flags(struct x86_mmu_pdpt *pdpt, void *ptr, size_t size,
+#ifdef CONFIG_X86_KPTI
+	/* MMU_ENTRY_RUNTIME_USER is a special flag; it indicates that the
+	 * region isn't acessible to user mode at boot, but may be granted
+	 * access later. The implications are that the PDPT/PD entries will
```
incorrect, it just ensures that the page directory/tables are created

nvlsianpu commented:
> NVS fs test is failing, not entirely sure how to fix, @nvlsianpu any tips?

Is this still failing?

andrewboie commented:
> 
> 
> > NVS fs test is failing, not entirely sure how to fix, @nvlsianpu any tips?
> 
> Is this still failing?

No, I figured it out.

andyross commented on arch/x86/core/ia32/x86_mmu.c:
```diff
@@ -344,6 +360,169 @@ void z_x86_mmu_set_flags(struct x86_mmu_pdpt *pdpt, void *ptr, size_t size,
+static inline bool is_within_system_ram(uintptr_t addr)
+{
+	return (addr >= DT_PHYS_RAM_ADDR) &&
+		(addr < (DT_PHYS_RAM_ADDR + (DT_RAM_SIZE * 1024U)));
```
Should probably do this in 64 bit math.  Physical RAM on PAE systems can routinely blow past 4G and result in an overflow here (and thus incorrect behavior) for "correct" devicetree settings.  We don't have to support using it, but we shouldn't break if devicetree authors (or EFI/e820/multiboot loaders, etc...) don't understand our limitations.

andrewboie replied:
There is no support in Zephyr for PAE itself (and there never will be), we only use PAE tables for the NX bit and nothing else.

All kinds of things will break if we are working with memory addresses that don't fit in a uintptr_t, not just here.

I'm gonna add a BUILD_ASSERT() that RAM is <= 4GB.

andrewboie replied:
> "correct" devicetree settings.

We don't need to do this.
DTS enforces already that these values fit, if I try to kick up QEMU's RAM size to be above 4GB I get a build error:

> Error: qemu_x86.dts.pre.tmp:31.21-43 Value out of range for 32-bit array element

I don't think we need to do anything here. We are never going to support PAE in the 32-bit port, we only use PAE tables for the NX bit.

andyross commented on arch/x86/core/ia32/x86_mmu.c:
```diff
@@ -344,6 +360,169 @@ void z_x86_mmu_set_flags(struct x86_mmu_pdpt *pdpt, void *ptr, size_t size,
+	 * does not have the user flag set, and this address falls outside
+	 * of system RAM, then don't bother generating any tables for it,
+	 * we will never need them later as memory domains are limited to
+	 * regions within system RAM.
```
Described like that, this sounds like it should be an __ASSERT() and not a silent ignore condition.  What's the circumstance where this can legitimately happen in practice?

andrewboie replied:
MMIO memory ranges. They don't need tables in the user page tables, only the kernel ones.

andyross commented on arch/x86/core/ia32/x86_mmu.c:
```diff
@@ -344,6 +360,169 @@ void z_x86_mmu_set_flags(struct x86_mmu_pdpt *pdpt, void *ptr, size_t size,
+	 * or modifies it.
+	 */
+	if ((flags & MMU_ENTRY_EXECUTE_DISABLE) == 0) {
+		pde->ignored2 = 1;
```
Style: given the subtlety, wouldn't be a bad idea to wrap the usage of ignored2 into an API like "set/get_page_partial_nx()" or something.

andyross commented on dts/x86/ia32.dtsi:
```diff
@@ -28,18 +28,17 @@
 
 
-	flash0: flash@1000 {
+	flash0: flash@100000 {
```
This all seems fine, low memory is a booby trap, in particular the 640k-1M region which has the MMIO mappings for legacy devices.  Note that real hardware does have (and require) this memory though, so it's likely we'll have to come back here at some point and put in some kind of mapping/allocation layer to allow us to get to it.  But loading the binary at 1M is IMHO absolutely the right choice.

andyross commented on arch/x86/core/ia32/crt0.S:
```diff
@@ -308,10 +308,8 @@ __csSet:
-	jmp	z_cstart
-
+	/* call C portion of kernel initialization and never return */
+	call	z_x86_prep_c
```
Love it.

andrewboie replied:
Eventually, we should nuke z_kernel_arch_init() for all arches

andyross reviewed (commented):
Some notes.  The overflow case on physical RAM size is the only thing I think is really needed.

Commit:
x86: add build assert that RAM bounds <= 4GB

DTS will fail this first, but there's no cost to adding
a second level of defense.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>
